tardis/visualization/tools/convergence_plot.py

```diff
@@ -0,0 +1,360 @@
+            legendgroup=f"group-{self.current_iteration}",
+            showlegend=False,
+            customdata=customdata,
+            hovertemplate="<b>Y</b>: %{y:.2f} at <b>X</b> = %{x:,.0f}%{customdata}",
+        )
+
+        # this adds dilution factor vs shell velocity plot
+        self.plasma_plot.add_scatter(
+            x=x,
+            y=self.iterable_data["w"],
+            line_color=self.colorscale[self.current_iteration - 1],
+            row=1,
+            col=2,
+            legendgroup=f"group-{self.current_iteration}",
+            name=self.current_iteration,
+            customdata=customdata,
+            hovertemplate="<b>Y</b>: %{y:.2f} at <b>X</b> = %{x:,.0f}%{customdata}",
+        )
+
+    def update_luminosity_plot(self):
+        """
+        Updates the plasma plots using the data collected using the fetch_data function.
+        This function is run every iteration.
+        """
+        x = list(range(1, self.iterations + 1))
+        with self.luminosity_plot.batch_update():
+            for index, luminosity in zip(range(3), self.luminosities):
+                self.luminosity_plot.data[index].x = x
+                self.luminosity_plot.data[index].y = self.value_data[luminosity]
+                self.luminosity_plot.data[index].hovertemplate = (
+                    "<b>%{y:.2g}</b>" + "<br>at X = %{x}<br>"
+                )
+
+                y = [
+                    ((emitted - requested) * 100) / requested
+                    for emitted, requested in zip(
+                        self.value_data["Emitted"], self.value_data["Requested"]
+                    )
+                ]
+                self.luminosity_plot.data[-2].x = x
+                self.luminosity_plot.data[-2].y = y
+                self.luminosity_plot.data[
+                    -2
+                ].hovertemplate = "Residual Luminosity: %{y:.2f}% at X = %{x:,.0f}<extra></extra>"
+
+                self.luminosity_plot.data[-1].x = x
+                self.luminosity_plot.data[-1].y = self.value_data["t_inner"]
+                self.luminosity_plot.data[
+                    -1
+                ].hovertemplate = "Inner Body Temperature: %{y:.2f} at X = %{x:,.0f}<extra></extra>"
+
+    def update(self, export_cplots=False, last=False):
+        """
+        Calls functions used to build and update convergence plots.
+        """
+        if self.iterable_data != {}:
+            if self.current_iteration == 1:
+                self.build()
+
+            self.update_plasma_plots()
+            self.update_luminosity_plot()
+            if last:
+                if hasattr(self, "plasma_plot_config"):
+                    if "data" in self.plasma_plot_config:
+                        self.override_plot_parameters(
+                            self.plasma_plot, self.plasma_plot_config
+                        )
+
+        self.current_iteration += 1
+        if export_cplots:
+            with suppress(TraitError):
```

Comment:
Good that you supressed only TraitError. Did you figure out that why this error arises (becuase supressing should be our last resort)?

Reply:
The TraitError arises because it was not accepting None as a return value, but `fig.show()` returns None.
```
TraitError: The 'children' trait of a VBox instance contains an Instance of a TypedTuple which expected a Widget, not the NoneType None.
```
No error arises when displaying the plots naturally:
```py 
display(widgets.VBox([go.FigureWidget(), go.FigureWidget()])) # No error
```
But only when one uses `fig.show()` to display the plots (we are doing this only when we are exporting plots)
```py
display(widgets.VBox([go.FigureWidget().show(), go.FigureWidget().show()])) # TraitError
```